sw/host/opentitanlib/src/util/file.rs

```diff
@@ -2,22 +2,74 @@
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
 
-use anyhow::Result;
+use anyhow::{anyhow, ensure, Result};
 use nix::libc::c_int;
 use nix::poll;
+use pem_rfc7468::{Decoder, Encoder, LineEnding};
+
 use std::convert::TryInto;
 use std::fs::File;
 use std::io::{self, Read, Write};
 use std::os::unix::io::{AsRawFd, RawFd};
 use std::path::Path;
 use std::time::Duration;
 
+/// Trait for data that can be written to and read from PEM files.
+pub trait PemSerilizable: ToWriter + FromReader {
+    /// The label for the PEM file.
+    ///
+    /// Appears around the base64 encoded data.
+    /// -----BEGIN MY_LABEL-----
+    /// ...
+    /// -----END MY_LABEL-----
+    fn label() -> &'static str;
+
+    /// Write to PEM file with label from `Self::label()`.
+    fn write_pem_file(&self, path: &Path) -> Result<()> {
+        const MAX_PEM_SIZE: usize = 4096;
+
+        let mut bytes = Vec::<u8>::new();
+        self.to_writer(&mut bytes)?;
+
+        let mut buf = [0u8; MAX_PEM_SIZE];
+        let mut encoder =
+            Encoder::new(Self::label(), LineEnding::LF, &mut buf).map_err(|e| anyhow!(e))?;
+        encoder.encode(&bytes).map_err(|e| anyhow!(e))?;
+        let len = encoder.finish().map_err(|e| anyhow!(e))?;
+
+        let mut file = File::create(path)?;
+        Ok(file.write_all(&buf[..len])?)
+    }
+
+    /// Read in from PEM file, ensuring the label matches `Self::label()`.
+    fn read_pem_file(path: &Path) -> Result<Self> {
+        let mut file = File::open(path)?;
+        let mut pem = Vec::<u8>::new();
+        let len = file.read_to_end(&mut pem)?;
+
+        let mut decoder = Decoder::new(&pem[..len]).map_err(|e| anyhow!(e))?;
+        ensure!(
+            decoder.type_label() == Self::label(),
+            "PEM labels do not match \"{}\" != \"{}\"",
+            decoder.type_label(),
+            Self::label()
+        );
```

Comment:
Can we create a formal error type for this condition?  You can use the `thiserror` crate (already in our deps):
```
use thiserror::Error;

#[derive(Debug, Error)]
enum MyError {
    #[error("PEM type error; expecting {0:?} but got {1:?}"]
    PemTypeError(String, String),
}

...
ensure!(a == b, MyError::PemTypeError(a, b));
```